# Day 1 Reading Journal

Hi everyone.  These are the exercises for the first set of readings from ThinkPython.  While this journal consists of several required exercises, it is meant to encourage active reading more generally.  You should use this journal to take detailed notes, catalog questions, and explore the content from ThinkPython deeply.

For more information on using the Jupyter notebook, check out the documentation [here](http://nbviewer.jupyter.org/github/jupyter/notebook/blob/master/docs/source/examples/Notebook/Examples%20and%20Tutorials%20Index.ipynb).  Additionally, in order to write text in this document, you will be making use of markdown cells. A handy cheatsheet for markdown is available [here](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet).

## Chapter 1

Note: the exercise numbers below match up with the reading for ease of cross referencing.

Python as well as C, C++, Perl, and Java are a high-level languages

1. Low-level languages are "machine languages" are are what computers run programs in
2. High-level languages are easier to program in, easier to read, and are portable, meaning they can run on different kinds of computers easily.

Interpreters and compilers process high-level languages into low-level

1. Interpreter reads high-level program and executes it a little at a time
2. compiler reads program and translates it before program starts running - the high-lelvel program is the source code and the translated program is the object code or executable

Python is an interpreted language and can be used in interactive mode and script mode

Program - sequence of instructions that specifies how to perform a computation

Basic instructions that appear in every language: input, output, math, conditional execution, repetition

1. Syntax errors - syntax refers to structure of a program
2. Runtime errors - are called exceptions and appears after the program has started running
3. Semantic errors - computer wont generate error message but it doesnt do the thing you wanted

Natural languages - evolved naturally
Formal languages - design by people, programming languages are formal

Parsing - figuring out the structure of a sentence 

In [1]:
print 'Hello, World!'

Hello, World!


In [2]:
print 'Hi

SyntaxError: EOL while scanning string literal (<ipython-input-2-d41356865299>, line 1)

### Exercise 3

Type `help()` to start the online help utility. Or you can type help('print') to get information about the print statement.  You should type `q` and then hit `enter` in the text box to exit the help utility. 

Note: this exercise is pretty simple (and there's not much to put in the box)!  We just want to make sure that you have tried out this super-handy feature of Python!

In [ ]:
help()


Welcome to Python 2.7!  This is the online help utility.

If this is your first time using Python, you should definitely check out
the tutorial on the Internet at http://docs.python.org/2.7/tutorial/.

Enter the name of any module, keyword, or topic to get help on writing
Python programs and using Python modules.  To quit this help utility and
return to the interpreter, just type "quit".

To get a list of available modules, keywords, or topics, type "modules",
"keywords", or "topics".  Each module also comes with a one-line summary
of what it does; to list the modules whose summaries contain a given word
such as "spam", type "modules spam".

help> print
The ``print`` statement
***********************

   print_stmt ::= "print" ([expression ("," expression)* [","]]
                  | ">>" expression [("," expression)+ [","]])

``print`` evaluates each expression in turn and writes the resulting
object to standard output (see below).  If an object is not a string,
it is first converted

### Exercise 4  

Start the Python interpreter and use it as a calculator. Python’s syntax for math operations is almost the same as standard mathematical notation. For example, the symbols +, - and / denote addition, subtraction and division, as you would expect. The symbol for multiplication is *.

If you run a 10 kilometer race in 43 minutes 30 seconds, what is your average time per mile? What is your average speed in miles per hour? (Hint: there are 1.61 kilometers in a mile). 

In [3]:
10/1.61

6.211180124223602

In [11]:
((43*60.)+30.)/3600.

0.725

In [14]:
(((43*60.)+30.)/60.)/6.211

7.003703107390114

In [12]:
6.211/0.725

8.566896551724138

Average time is 7 minutes per mile and average speed is 8.57 miles per hour. 

## Chapter 2

type() tells what type a value has
Types include strings, integers, and floating-points or floats which are decimals

Variables - names that refers to values
Assignment statements create new variables and gives them values

keywords such as class cannot be used as variable names

Operators - special symbols that represent computations
Python performs floor division

Expression - combination of values, variables, and operators

Statement - unit of code that the Python interpreter can execute

Order of evaluation depends on rules of precedence
PEMDAS - parentheses, exponentiation, multiplication and division, addition and subtraction

+ operator performs concatenation which links the strings end-to-end
* operator performs repitition on strings

comments begin with the # symbol

### Exercise 2  

Assume that we execute the following assignment statements:

width = 17
height = 12.0
delimiter = '.'

For each of the following expressions, write the value of the expression and the type (of the value of the expression).

1. `width/2`
2. `width/2.0`
3. `height/3`
4. `1 + 2 * 5`
5. `delimiter * 5`

Please use the following markdown cell to provide your answer.

1. 8, int
2. 8.5, float
3. 4.0, float
4. 11, int
5. '.....', str



### Exercise 3  

Practice using the Python interpreter as a calculator.  For each of these provide Python code that computes each answer.

    


The volume of a sphere with radius r is 4/3 $\pi r^3$. What is the volume of a sphere with radius 5? Hint: 392.7 is wrong!

In [4]:
r = 5.0
pi = 3.1415
(4.0/3.0)*pi*(r**3)

523.5833333333333

Suppose the cover price of a book is \$24.95, but bookstores get a 40% discount. Shipping costs \$3 for the first copy and 75 cents for each additional copy. What is the total wholesale cost for 60 copies?

In [6]:
(60*(24.95*0.6))+(3+(0.75*59))

945.4499999999999

If I leave my house at 6:52 am and run 1 mile at an easy pace (8:15 per mile), then 3 miles at tempo (7:12 per mile) and 1 mile at easy pace again, what time do I get home for breakfast? 

In [5]:
(8.25*2)+(7.2*3)

38.1

You get home at 7:30 am.

## Chapter 3

function - named sequence of statements that performs a computation
int, str, and float are functions

In [6]:
int('21')

21

module - file that contains a collection of related functions
have to import before using a module

left side of an assignment statement has to be a variable name
eg. minutes = hours * 60

function definition - specifies name of new function and sequence of statements that execute when the function is called

e.g.
def print_lyrics():
    print "I'm a lumberjack..."
    
print_lyrics is the name of the function

flow of execution - the order in which statements are executed

statements inside the function are not executed until function is called - therefore, makes more sense to follow flow of execution and not read a program from top to bottom

In [7]:
def print_twice(a):
    print a
    print a
    
print_twice('Hello!')

Hello!
Hello!


Variables inside functions are local, meaning it only exists in the function

In [8]:
def cat_twice(a,b):
    cat = a+b
    
print cat

NameError: name 'cat' is not defined

stack diagrams - show value of each variable and show the function each variable belongs to
frame - represents a function in a stack diagram

fruitful functions yield results while void functions just perform an action

In [10]:
import math
sin(pi)

NameError: name 'sin' is not defined

In [9]:
from math import *
print pi
sin(pi/2)

3.14159265359


1.0

### Exercise 3

Python provides a built-in function called len that returns the length of a string, so the value of len('allen') is 5.
Write a function named right_justify that takes a string named s as a parameter and prints the string with enough leading spaces so that the last letter of the string is in column 70 of the display.

```
>>> right_justify('allen')
                                                                 allen```

In [2]:
def right_justify(s):
    space = ' '
    space_length = 70-len(s)
    print space*space_length + s
right_justify('allen')

                                                                 allen


### Exercise 5

This exercise can be done using only the statements and other features we have learned so far.

(a) Write a function that draws a grid like the following:
```
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
```
**Hint:** to print more than one value on a line, you can print a comma-separated sequence:
print '+', '-'
If the sequence ends with a comma, Python leaves the line unfinished, so the value printed next appears on the same line.
print '+', 
print '-'
The output of these statements is '+ -'.
A print statement all by itself ends the current line and goes to the next line.

In [2]:
def plus():
    print '+',
def minus():
    print '-',
def line():
    print '|',
def space():
    print ' ',
def four(a):
    a()
    a()
    a()
    a()
def row_one():
    plus()
    four(minus)
    plus()
    four(minus)
    plus()
    print
def row_two():
    line()
    four(space)
    line()
    four(space)
    line()
    print
    
def grid():
    row_one()
    four(row_two)
    row_one()
    four(row_two)
    row_one()
    
grid()

+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +
|         |         |
|         |         |
|         |         |
|         |         |
+ - - - - + - - - - +


(b) Write a function that draws a similar grid with four rows and four columns.

In [3]:
def row(a,b):
    a()
    four(b)
    a()
    four(b)
    a()
    four(b)
    a()
    four(b)
    a()
    print
def row_one():
    row(plus,minus)
def row_two():
    row(line,space)
def grid():
    row(row_one,row_two)
grid()

+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
|         |         |         |         |
+ - - - - + - - - - + - - - - + - - - - +



## Chapter 5.1 - 5.7

Modulus operator yields the remainder of division and is the percent sign (%)
Also x % 10 yields the right-most digid of x and x % 100 yeilds the last 2 digits

boolean expression - expression that is true or false
True and False are type bool
some relational operators inclue ==, !=, >, <, >=, <=

and, or, and not are the three logical operators

In [14]:
a = 1
b = 2
not (a<b)

False

Conditional statements give ability to check conditions and change behavior of program accordingly

Alternative execution is a form of the if statement that has two possibilities

Chained conditionals have more than two possiblities

nested conditionals are conditionals within each other

In [19]:
x = 23
if x%2 == 0:
    print 'x is even'
else:
    print 'x is odd'

x is odd


In [21]:
if a < 10:
    if b<5:
        print 'a*b must be less than 40'

a*b must be less than 40


### Exercise 3  

Fermat’s Last Theorem says that there are no positive integers a, b, and c such that $a^n + b^n = c^n$ for any values of n greater than 2.

(a) Write a function named `check_fermat` that takes four parameters—a, b, c and n—and that checks to see if Fermat’s theorem holds. If n is greater than 2 and it turns out to be true that
$a^n + b^n = c^n$ the program should print, "Holy smokes, Fermat was wrong!" Otherwise the program should print, "No, that doesn’t work."

In [27]:
def check_fermat (a,b,c,n):
    if n > 2:
        if a**n + b**n == c**n:
            print "Holy smokes, Fermat was wrong!"
        else:
            print "No, that doesn't work."
            
check_fermat (1,2,3,12)

No, that doesn't work.


(b) Write a function that prompts the user to input values for a, b, c and n, converts them to integers, and uses check_fermat to check whether they violate Fermat’s theorem.

In [46]:
def prompt_fermat():
    a = raw_input('Enter a number for a\n')
    a = int(a)
    b = raw_input('Enter a number for b\n')
    b = int(b)
    c = raw_input('Enter a number for c\n')
    c = int(c)
    n = raw_input('Enter a number for n\n')
    n = int(n)
    check_fermat (a,b,c,n)
prompt_fermat()

Enter a number for a
1
Enter a number for b
2
Enter a number for c
3
Enter a number for n
4
No, that doesn't work.


### Exercise 4  

If you are given three sticks, you may or may not be able to arrange them in a triangle. For example, if one of the sticks is 12 inches long and the other two are one inch long, it is clear that you will not be able to get the short sticks to meet in the middle. For any three lengths, there is a simple test to see if it is possible to form a triangle:
> If any of the three lengths is greater than the sum of the other two, then you cannot form a triangle. Otherwise, you can. (If the sum of two lengths equals the third, they form what is called a “degenerate” triangle.)

(a) Write a function named `is_triangle` that takes three integers as arguments, and that prints either "Yes" or "No," depending on whether you can or cannot form a triangle from sticks with the given lengths.

In [36]:
def is_triangle(a,b,c):
    if c>a+b:
        print "No"
    elif b>a+c:
        print "No"
    elif a>b+c:
        print "No"
    else:
        print "Yes"
        
is_triangle(1,3,5)

is_triangle (10,3,1)

is_triangle (2, 12, 9)

No
No
No


(b) Write a function that prompts the user to input three stick lengths, converts them to integers, and uses is_triangle to check whether sticks with the given lengths can form a triangle.

In [47]:
def triangle_prompt():
    a = raw_input('Enter a length for a\n')
    a = int(a)
    b = raw_input('Enter a length for b\n')
    b = int(b)
    c = raw_input('Enter a length for c\n')
    c = int(c)
    is_triangle(a,b,c)
triangle_prompt()

Enter a length for a
1
Enter a length for b
3
Enter a length for c
5
No


## Notes for the Instructors

Please use the space below to write comments to help us plan the next class session.  For instance, if you want to see us go over an example of a particular type of problem, you can indicate that here.

Please remember that the best way to get quick feedback from the instructors as well as your peers, is to use Piazza.  However, consider using this channel if it doesn't make sense to post your note to Piazza.